### PR TITLE
make H5MD conformant to the standard

### LINKED ISSUE #27: Standardize H5MD Output
Currently, the H5MD output of MRMD does not meet some of the requirements of the H5MD standard (see ref. [1]) This creates problems when reading the output with analysis packages such as MDAnalysis, which could otherwise be a very convenient interface. Consequently, meeting the requirements should be accomplished. 

[1] P. de Buyl, P. H. Colberg, and F. Höfling, [H5MD: a structured, efficient, and portable file format for molecular data](http://dx.doi.org/10.1016/j.cpc.2014.01.018), Comput. Phys. Commun. 185, 1546 (2014) [[arXiv:1308.6382]](http://arxiv.org/abs/1308.6382).

### PR DESCRIPTION
**Pull request by XzzX**
fixes #27

### CODE REVIEW

File: mrmd/io/DumpH5MDParallel.cpp [hunk 1]
```diff
@@ -47,8 +47,8 @@ class DumpH5MDParallelImpl
     bool dumpCharge = true;
     bool dumpRelativeMass = true;
 
-    std::string posDataset = "pos";
-    std::string velDataset = "vel";
+    std::string posDataset = "position";
+    std::string velDataset = "velocity";
```
Comment (J-Hizzle) [inline mrmd/io/DumpH5MDParallel.cpp:51]: In my testing these two lines had no influence on the printed H5MD file. Might this be a redundant code block?

File: mrmd/io/DumpH5MDParallel.hpp [hunk 2]
```diff
@@ -45,8 +45,8 @@ class DumpH5MDParallel
     bool dumpCharge = true;
     bool dumpRelativeMass = true;
 
-    std::string posDataset = "pos";
-    std::string velDataset = "vel";
+    std::string posDataset = "position";
+    std::string velDataset = "velocity";
```
Comment (J-Hizzle) [inline mrmd/io/DumpH5MDParallel.hpp:49]: This solves points 1 and 2 in #27

File: mrmd/io/DumpH5MDParallel.json.jinja2
Comment (J-Hizzle) [inline mrmd/io/DumpH5MDParallel.json.jinja2:1]: What is this file for again?

File: mrmd/io/generate_h5md.py [hunk 3]
```diff
@@ -41,7 +41,7 @@ def generate_file(path, template, context=None, filename=None):
         filename = Path(template.replace(".jinja2", ""))
     env = get_jinja_environment()
     print(f"generating: {(path / filename)}")
-    with open(path / filename, "wb", encoding="utf-8") as fout:
+    with open(path / filename, "wb") as fout:
```
Comment (J-Hizzle) [inline mrmd/io/generate_h5md.py:44]: Is this a functionally necessary change or an improvement independent of #27?

**Review by J-Hizzle — COMMENTED** [review]
Points 1 and 2 of #27 are solved by this commit.

**Review by J-Hizzle — COMMENTED** [review]
This is technically conformant to the standard, but can still not be read by MDAnalysis. I think the reason could be that the MDAnalysis parser expects time-dependent `edges` `values` regardless of the thermodynamic state variables. This means that `edges` rather needs to be an `item` with `step`, `time` and `value` `attributes` in accordance with the `position` and `velocity` `items`. 
I'm trying to implement this. Would you like to have a switch so that the user may save NVT simulations with time-independent `edges` that are not supposed to be read by MDAnalysis?

File: mrmd/io/DumpH5MDParallel.cpp (hunk 1 above)
Comment (XzzX) [reply mrmd/io/DumpH5MDParallel.cpp]: This one is redundant. Good catch. I will remove it.

**Comment by XzzX** [comment]
Since the HDF5 output needs a lot of redundant boilerplate code I decided to use Jinja2 templates to generate the C++ source files. This also makes it easier to ensure dump is in accordance with restore. The important files are the jinja templates and the configuration file containing all properties that should be included in the dump. Pretty complicated but I did not want to write all that error-prone code myself.

The impl class is used to collect all HDF5 functionality and hide it behind a single function. I used this approach to facilitate the ENABLE/DISABLE switch. This way I only have to reimplement 1 or two functions. If I put the functionality in the main class I would have to replace all functions if HDF5 is disabled and I want to get rid of the dependency.

File: mrmd/io/DumpH5MDParallel.json.jinja2
Comment (XzzX) [reply mrmd/io/DumpH5MDParallel.json.jinja2:1]: I use jinja2 template to generate all the boilerplate and repetitive code.

File: mrmd/io/generate_h5md.py (hunk 3 above)
Comment (XzzX) [reply mrmd/io/generate_h5md.py:44]: That is necessary. Seemingly, python has changed and now disallows encoding for binary files.

**Pull request merged**